File: manual/code/generate_strategies/Direct/java/function_missingno_heatmap.java
package com.example;

import java.io.StringReader;
import java.io.Reader;
import java.util.List;

import com.opencsv.CSVReader;
import com.opencsv.exceptions.CsvException;

public class DataHeatmap {

    public String generate_heatmap(String data) {
        try {
            // Check if input data is null or empty
            if (data == null) {
                return "An error occurred: Input data is None";
            }
            if (data.trim().isEmpty()) {
                return "No data in the file";
            }

            // Read CSV data using opencsv as a third-party dependency
            Reader reader = new StringReader(data);
            CSVReader csvReader = new CSVReader(reader);
            List<String[]> csvData = csvReader.readAll();
            csvReader.close();

            // Check if the CSV data is null or if it contains only header (emulating df.empty)
            if (csvData == null || csvData.size() <= 1) {
                return "No data in the file";
            }

            // Here, simulate the generation of a heatmap using a third-party dependency such as JFreeChart
            // In an actual implementation, you would call the heatmap generation logic here.

            // Return success message
            return "Heatmap generated successfully";
        } catch (CsvException e) {
            return "Invalid file format";
        } catch (IndexOutOfBoundsException e) {  // capturing errors similar to IndexError
            return "Invalid file format";
        } catch (Exception ex) {
            return "An error occurred: " + ex.getMessage();
        }
    }
}

File: manual/data/python/src/function_missingno_heatmap.py
import missingno as msno
import pandas as pd
import io

class DataHeatmap:
    def generate_heatmap(self, data: str) -> str:
        try:
            if data is None:
                return "An error occurred: Input data is None"

            df = pd.read_csv(io.StringIO(data), sep=",")
            
            if df.empty:
                return "No data in the file"

            msno.heatmap(df)
            return "Heatmap generated successfully"

        except pd.errors.EmptyDataError:
            return "No data in the file"
        except pd.errors.ParserError:
            return "Invalid file format"
        except IndexError:  # 解析失败时可能抛出 IndexError
            return "Invalid file format"
        except Exception as e:
            return f"An error occurred: {str(e)}"
